### ISSUE TITLE
Incorrect transformer mask size

### ISSUE THREAD

**Issue by egaznep** [speechbrain/speechbrain#2344]
### Describe the bug

PyTorch `multi_head_attention` code enforces that data size and mask size matches.

https://github.com/pytorch/pytorch/blob/df4e3d9d08f3d5d5439c3626be4bf29659488cdf/torch/nn/functional.py#L5442-L5444

However, `speechbrain` code generates masks according to the longest `wav_len`, which can be shorter than the batch size with padding, resulting in the exception given in the codeblock above.
https://github.com/speechbrain/speechbrain/blob/f6e297e65e266b5b1b8dbbfbf9d60520db9cfd88/speechbrain/lobes/models/transformer/TransformerASR.py#L146

Solution:

The function `length_to_mask` accepts an optional argument `max_len`, this could be used. Should I open a PR?

https://github.com/speechbrain/speechbrain/blob/f6e297e65e266b5b1b8dbbfbf9d60520db9cfd88/speechbrain/dataio/dataio.py#L758-L772

### Expected behaviour

The mask should have been generated according to what PyTorch anticipates. Instead, an exception is thrown.

### Environment Details

```
speechbrain==0.5.16  (checking the develop branch this still seem to be an issue)
pytorch==2.1.1
```

**Comment by Adel-Moumen** [comment]
Hello @egaznep, thanks for opening this issue! 

Could you please have a look @TParcollet? Thanks :)

**Comment by TParcollet** [comment]
Hello @egaznep I am not sure to understand the issue here. Could you provide a code snippet showing explicitly the error? The function length_to_mask() is expected to provide masks containing padding i.e. real size of the input tensor, and this for each sequence. Could you detail a bit more what you are trying to achieve?

**Comment by egaznep** [comment]
Here is a minimal working (or in this case crashing) example:

```python
import torch
import torch.nn as nn
from speechbrain.lobes.models.transformer.TransformerASR import TransformerASR
import random


# Instantiate the TransformerASR model
model = TransformerASR(
    tgt_vocab=720,
    input_size=80,
    d_model=512,
    nhead=1,
    num_encoder_layers=1,
    num_decoder_layers=1,
)

# Generate some dummy input with different lengths
input_lengths = torch.tensor([l for l in range(10,101,10)])
input_data = [torch.randn(length,80) for length in input_lengths]
input_targets = [torch.randint(low=0, high=720, size=(length.item(),)) for length in input_lengths]
# Pad the input sequences to have the same length
input_data = nn.utils.rnn.pad_sequence(input_data, batch_first=True)
input_targets = nn.utils.rnn.pad_sequence(input_targets, batch_first=True)
input_lengths = input_lengths/100.0

print(input_data.shape, input_targets.shape, input_lengths.shape)

output = model.forward(input_data, input_targets, wav_len=input_lengths) # works
output = model.forward(input_data[:-1], input_targets[:-1], wav_len=input_lengths[:-1]) # fails
```

First call to the model.forward in line 29 has the following `wav_len`s: 10, 20, ..., 100, thus all are padded to 100. The mask computation will not fail because it can infer the proper mask size.

Second call to the model.forward in line 30, however, has the following `wav_len`s: 10, 20, ..., 90, even though all are padded to 100 because that is the longest in the entire dataset. The mask computation will fail because mask generator does not check how long the padded sequences are, but simply the largest `wav_len`s are.

**Comment by TParcollet** [comment]
Hello thanks. SpeechBrain padding is relative to the batch, not the dataset. The max len of wav_lens is the max len of the batch.

**Comment by egaznep** [comment]
I had this error while training a model using `DistributedDataParallel` on 2 GPUs. Could it be that initial data sampler does the padding relative to the complete batch size, but then during distribution to individual GPUs the error I am facing occurs?

**Comment by TParcollet** [comment]
@Gastron correct me if I am wrong, but as far as I know, DDP sampler is per-process, hence the padding should be relative to that batch. @egaznep any chance that you could give us an example where this happens?

**Comment by Adel-Moumen** [comment]
Hello @egaznep, any news on your side please ?
